cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -708,9 +708,16 @@ void MapBuilderBridge::ProcessTSDFMesh(pcl::PolygonMesh &mesh,
     const auto submap3d =
         static_cast<const ::cartographer::mapping::Submap3D *>(
             data.begin()->data.submap.get());
-    const auto tsdf =
-        static_cast<const ::cartographer::mapping::HybridGridTSDF *>(
-            &submap3d->high_resolution_hybrid_grid());
+    const ::cartographer::mapping::HybridGridTSDF *tsdf;
+    if (kTsdfVisualizationHighRes) {
+      tsdf =
+          dynamic_cast<const ::cartographer::mapping::HybridGridTSDF *>(
+              &submap3d->high_resolution_hybrid_grid());
+    } else {
+      tsdf =
+          dynamic_cast<const ::cartographer::mapping::HybridGridTSDF *>(
+              &submap3d->low_resolution_hybrid_grid());
+    }
     pcl::PointCloud<pcl::PointXYZ> cloud;
     auto local_trajectory_data = GetLocalTrajectoryData();
```

Comment:
As you only need read access, `auto const &` instead of `auto` should be sufficient and avoid the copy.